include/hc.hpp

```diff
@@ -1305,13 +1305,15 @@ class completion_future {
      *                     the expense of using one CPU core for active waiting.
      */
     void wait(hcWaitMode mode = hcWaitModeBlocked) const {
-        if (this->valid()) {
-            if (__asyncOp != nullptr) {
-                __asyncOp->setWaitMode(mode);
-            }   
-            //TODO-ASYNC - need to reclaim older AsyncOps here.
+        if (__amp_future.valid()) {
+            if (__asyncOp != nullptr) throw runtime_exception("completion_future expected amp, had async op", 0);
             __amp_future.wait();
         }
+        if (__asyncOp != nullptr) {
+            if (__amp_future.valid()) throw runtime_exception("completion_future expected async op, had amp", 0);
```

Comment:
Aren't new lines 1308-1309 and 1312-1313 identical in functionality unless the order of the conditions checked matter? If so, some comment would be helpful.
If not, would it be ok to factor it out as: 
`if (__amp_future.valid() &&  (__asyncOp != nullptr)) {`
`    throw runtime_exception("completion_future expected amp, had async op", 0);`
`}`

It could reduce indent of main code.